### PR TITLE
Upgradeability to a new Asset Pool

### PR DESCRIPTION
**Pull request by dimpar**
Refs: #74

Governance can approve a new Asset Pool from the CoveragePool contract. Underwriter will be able to move their collateral tokens from the old asset pool to a new one. Old underwriter tokens will be burned and new ones will be calculated, minted and then transferred back to an underwriter in exchange for their collateral tokens.

### CODE REVIEW

File: contracts/AssetPool.sol
Comment (pdyraga) [inline contracts/AssetPool.sol:304]: Our token implementation has only one owner that is allowed to mint it. We can rework the token to allow for multiple minters but from a design perspective, it feels better if there was only one minter for a token.

Comment (pdyraga) [inline contracts/AssetPool.sol:304]: I do not think we can assume COV amount minted by the second pool should be the same. It could happen that there was a claim and the math may not hold in this case.

File: contracts/AssetPool.sol
Comment (pdyraga) [inline contracts/AssetPool.sol:301]: I am not sure about it.... If we just delete them, and there is a pending withdrawal, the underwriter will lose a part of their tokens.

Comment (dimpar) [reply contracts/AssetPool.sol:301]: Wouldn't [this](https://github.com/keep-network/coverage-pools/blob/upgradeability-for-asset-pool/contracts/AssetPool.sol#L274) line do the trick? I take that we're interested in moving the entire balance of underwriter to a new Asset Pool. Can't we read the total balance no matter if it's pending or not and then transfer it?

File: contracts/AssetPool.sol
Comment (dimpar) [reply contracts/AssetPool.sol:304]: Okay, I take that the problem will pop up with the old AssetPool being an owner of `newUnderwriterToken` and a new AssetPool won't be able to mint `newUnderwriterTokens`? If yes, wdyt about transferring an ownership to a new AssetPool right after we mint new tokens, ie. `newUnderwriterToken.transferOwnership(address(newAssetPool))`?

Comment (dimpar) [reply contracts/AssetPool.sol:304]: Do you mean that a claim could happen a transaction early? If yes, then a couple of lines above in this code we have the following, which reads the latest balance of the collateral token.
```
rewardsPool.withdraw(); 
uint256 collateralBalance = collateralToken.balanceOf(address(this));
```

**Comment by pdyraga** [comment]
Answering the three ([1](https://github.com/keep-network/coverage-pools/pull/76#discussion_r645367225), [2](https://github.com/keep-network/coverage-pools/pull/76#discussion_r645372062), [3](https://github.com/keep-network/coverage-pools/pull/76#discussion_r645355202)) comments at once:
- I think only the given `AssetPool` should be eligible to mint `UnderwriterToken` and we should always have 1<>1 relationship between the pool and the token.
- If someone is exiting from the pool, they need to be able to complete their withdrawal. We can optionally let them cancel the withdrawal and migrate to the new pool but this case will be so rare that it does not justify the complexity we'd have to introduce to the contract.
- COV amounts between pools might not match. It is up to the new pool to decide about the amount of COV that should be minted based on the collateral transferred and the current pool's balance.

Here are the building blocks I think we need (disclaimer: I haven't had a chance to validate it in the code):
- `approveNewPoolUpgrade(address newAssetPool)` for the governance
- `upgradeToNewPool(uint256 covAmount)` for the underwriter - needs to transfer COV from the underwriter, compute the collateral value and send the collateral to the new pool
- `deposit(uint256 amount, address underwriter)` - just like the standard deposit except that mints COV to `underwriter` and not `msg.sender`.

**Comment by dimpar** [comment]
@pdyraga can you pls take a look again?

File: contracts/AssetPool.sol
Comment (pdyraga) [inline contracts/AssetPool.sol:28]: We could have a `depositFor(uint256 amount, address covRecipient)` function. This would simplify the design - we wouldn't have to compute the balance of new COV when updating and we wouldn't need this reference.

File: contracts/AssetPool.sol
Comment (pdyraga) [inline contracts/AssetPool.sol:73]: Maybe just `AssetPoolUpgraded`? This is just a one-step process, so there is no initiation & completion.

File: contracts/AssetPool.sol
Comment (pdyraga) [inline contracts/AssetPool.sol:76]: I think this should be the COV amount from this pool. This is what `WithdrawalInitiated` includes and this is the event from the current contract, so should inform about the current contract's state changes.

File: contracts/AssetPool.sol
Comment (pdyraga) [inline contracts/AssetPool.sol:351]: Worth emitting an event.

File: contracts/AssetPool.sol
Comment (pdyraga) [inline contracts/AssetPool.sol:340]: The owner of this contract is the coverage pool so this will not work unless the idea was to route this call through `CoveragePool` which I think is acceptable and probably the best solution here.

File: contracts/interfaces/IAssetPool.sol
Comment (pdyraga) [inline contracts/interfaces/IAssetPool.sol:10]: I would also add `initiateWithdrawal` and `completeWithdrawal` to this interface. Withdrawal will always be a two-step process and even if not, only deposit function is needed for the upgrade.

File: contracts/AssetPool.sol
Comment (pdyraga) [reply contracts/AssetPool.sol]: The link is pointing now to a solhint line so I am not sure what was there :) But I think we should keep upgrade and withdrawal separate for simplicity.

File: contracts/AssetPool.sol
Comment (pdyraga) [reply contracts/AssetPool.sol]: I do not think this helps. New asset pool and old asset pool have their own COV supplies and collateral token balances.

File: contracts/AssetPool.sol
Comment (dimpar) [reply contracts/AssetPool.sol:28]: I think that was the initial approach, but based on [this](https://www.flowdock.com/app/cardforcoin/tech/threads/kLsWXXDV8tcJ2uXtKBQqxqDAUPA) conversation, you mentioned we won't need to pass underwriter and that's why we need `UnderwriterToken` here to make it happen.

Okay, I'll change it to the initial approach.

File: contracts/AssetPool.sol
Comment (dimpar) [reply contracts/AssetPool.sol:76]: The reason behind it was that this contract was transferring new underwriter tokens to underwriters, but since the latest approach is that a new asset pool will handle transfer of new cov tokens we can change it

File: contracts/AssetPool.sol
Comment (dimpar) [reply contracts/AssetPool.sol:340]: Yup, that's the idea, will add this change in the next commits.

File: contracts/interfaces/IAssetPool.sol
Comment (dimpar) [reply contracts/interfaces/IAssetPool.sol:10]: Done

File: contracts/AssetPool.sol
Comment (dimpar) [reply contracts/AssetPool.sol]: Done

File: contracts/AssetPool.sol
Comment (dimpar) [reply contracts/AssetPool.sol]: Done

File: contracts/AssetPool.sol
Comment (dimpar) [reply contracts/AssetPool.sol]: yea, that comment was not relevant anymore.

File: contracts/AssetPool.sol
Comment (pdyraga) [inline contracts/AssetPool.sol:45]: This is no longer needed, right?

Comment (dimpar) [reply contracts/AssetPool.sol:45]: [here](https://github.com/keep-network/coverage-pools/blob/upgradeability-for-asset-pool/contracts/AssetPool.sol#L322) we call `depositFor()` on a new Asset Pool. Before we can call it, the governance has to [approve](https://github.com/keep-network/coverage-pools/blob/upgradeability-for-asset-pool/contracts/AssetPool.sol#L347) it. We need `newAssetPool` if we do it this way.

Comment (pdyraga) [reply contracts/AssetPool.sol:45]: Oops I am sorry, GitHub tricked me, I probably did not refresh the page before posting this comment. This comment was for line 41 with `newUnderwritertToken`.

Comment (dimpar) [reply contracts/AssetPool.sol:44]: Ah, right `newUnderwriterToken` is not needed. Removed f757e4622d79e20c4e7c1e07c1530d407fab334a

File: test/AssetPool.test.js
Comment (pdyraga) [inline test/AssetPool.test.js:1109]: Shouldn't `amountToUpgrade` be approved instead?

File: test/AssetPool.test.js
Comment (pdyraga) [inline test/AssetPool.test.js:1159]: Shouldn't `amountToUpgrade` be approved instead?

File: test/AssetPool.test.js
Comment (pdyraga) [inline test/AssetPool.test.js:1198]: Shouldn't `amountToUpgrade` be approved instead? Also, shouldn't we approve after one day passed. Should not do any difference technically but this is what will happen in the real world.

File: contracts/AssetPool.sol
Comment (dimpar) [reply contracts/AssetPool.sol]: e6dc71d

File: contracts/interfaces/IAssetPoolUpgrade.sol
Comment (dimpar) [reply contracts/interfaces/IAssetPoolUpgrade.sol]: But at the same time we are interested in a new Asset Pool conforming to `IAssetPool` as the old one does. Devs will have to remember to add `IAssetPool` to `NewAssetPool is IAssetPoolUpgrade, IAssetPool` cause it won't be caught at compile time. 
Having 3 empty functions in a test stub is no biggy imo, just shows what functions a real contract has to have. But anyway, I've separated them per your comment if you feel they should be separated. a6f7056

File: test/AssetPool.test.js
Comment (dimpar) [reply test/AssetPool.test.js]: 18fed51

File: test/AssetPool.test.js
Comment (dimpar) [reply test/AssetPool.test.js]: 18fed51

File: test/AssetPool.test.js
Comment (dimpar) [reply test/AssetPool.test.js]: 18fed51

File: test/AssetPool.test.js
Comment (dimpar) [reply test/AssetPool.test.js:1178]: 18fed51

File: test/AssetPool.test.js
Comment (dimpar) [reply test/AssetPool.test.js]: 18fed51

File: test/AssetPool.test.js
Comment (dimpar) [reply test/AssetPool.test.js]: 18fed51

File: test/AssetPool.test.js
Comment (dimpar) [reply test/AssetPool.test.js]: 18fed51

File: contracts/CoveragePool.sol
Comment (dimpar) [reply contracts/CoveragePool.sol:190]: Hm, not sure what was changed, but it's not complaining when I run it.

File: test/AssetPool.test.js
Comment (pdyraga) [inline test/AssetPool.test.js:1057]: Can we say instead "when new asset pool address does not match"? This could be a user error as well. Or, in extreme circumstances, upgrade of an upgrade conflicting with a pending underwriter's transaction.

Comment (pdyraga) [reply test/AssetPool.test.js:1057]: Actually... we can leave it as is, after reading the test one more time I see what was the reasoning.

File: test/AssetPool.test.js
Comment (pdyraga) [reply test/AssetPool.test.js]: > Also, shouldn't we approve after one day passed. Should not do any difference technically but this is what will happen in the real world.

This still needs to be addressed. For:
- when there was a claim
- when rewards were allocated

File: test/AssetPool.test.js
Comment (dimpar) [reply test/AssetPool.test.js]: fcab985 Agree on the latter, but I took that the whole point for adding this check in contract was because something bad can happen with the governance, hence the described scenerio context.

File: test/AssetPool.test.js
Comment (dimpar) [reply test/AssetPool.test.js]: Done in 089f488

File: test/AssetPool.test.js
Comment (pdyraga) [inline test/AssetPool.test.js:12]: `only` 😬

Comment (dimpar) [reply test/AssetPool.test.js]: oops.. removed

File: test/AssetPool.test.js
Comment (pdyraga) [inline test/AssetPool.test.js:197]: This is not needed here, or is it?

Comment (dimpar) [reply test/AssetPool.test.js]: eh, not sure how it got there... 83048dd

**Review by pdyraga — APPROVED** [review]

**Pull request merged**
